### PR TITLE
make kill on JobContextService async

### PR DESCRIPTION
**Pull request by dobe**

### CODE REVIEW

**Comment by dobe** [comment]
@matriv pls review

**Comment by dobe** [comment]
or @mfussenegger ?

File: sql/src/main/java/io/crate/executor/transport/kill/TransportKillJobsNodeAction.java [hunk 1]
```diff
@@ -21,71 +21,33 @@
...
-        this.jobContextService = jobContextService;
-        this.clusterService = clusterService;
-        this.transportService = transportService;
-
-        transportService.registerRequestHandler(TRANSPORT_ACTION,
-            KillJobsRequest.class,
-            ThreadPool.Names.GENERIC,
-            new NodeActionRequestHandler<KillJobsRequest, KillResponse>(this) {});
+        super("crate/sql/kill_jobs", settings, jobContextService, clusterService, transportService);
+        //register(KillJobsRequest.class);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/kill/TransportKillJobsNodeAction.java]: remove please

File: sql/src/test/java/io/crate/jobs/JobContextServiceTest.java [hunk 2]
```diff
@@ -129,7 +129,7 @@ public void innerKill(@Nonnull Throwable throwable) {
         @SuppressWarnings("unchecked")
         Map<UUID, JobExecutionContext> activeContexts = (Map<UUID, JobExecutionContext>) activeContextsField.get(jobContextService);
         assertThat(activeContexts.size(), is(1));
-        assertThat(jobContextService.killAll(), is(1L));
+        assertThat(jobContextService.killAll().get(), is(1));
```
Comment (mfussenegger) [inline sql/src/test/java/io/crate/jobs/JobContextServiceTest.java]: please add timeouts to the .get calls.

File: sql/src/main/java/io/crate/jobs/JobContextService.java [hunk 3]
```diff
@@ -134,47 +125,66 @@ public JobExecutionContext createContext(JobExecutionContext.Builder contextBuil
...
+        }
+        Collection<UUID> toKill = ImmutableList.copyOf(activeContexts.keySet());
+        if (toKill.isEmpty()) {
+            return Futures.immediateFuture(0);
+        }
+        return killContexts(toKill);
+    }
+
+    private ListenableFuture<Integer> killContexts(Collection<UUID> toKill) {
+        assert !toKill.isEmpty();
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/jobs/JobContextService.java]: please add an assertion message

File: sql/src/main/java/io/crate/executor/transport/kill/TransportKillNodeAction.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportRequest;
+import org.elasticsearch.transport.TransportService;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.concurrent.Callable;
+
+abstract class TransportKillNodeAction<Request extends TransportRequest> extends AbstractComponent
+    implements NodeAction<Request, KillResponse>, Callable<Request> {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/executor/transport/kill/TransportKillNodeAction.java]: Think I'm missing something - where is this used as Callable?

Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/kill/TransportKillNodeAction.java]: in the registration

**Comment by dobe** [comment]
@mfussenegger fixup pushed

File: sql/src/main/java/io/crate/executor/transport/kill/TransportKillNodeAction.java (hunk 4 above)
Comment (dobe) [inline sql/src/main/java/io/crate/executor/transport/kill/TransportKillNodeAction.java]: see https://github.com/crate/crate/pull/4213/files#diff-5ed3f966c168c06a9b1b4081a75d6b3bR66

**Pull request merged**